Add FORCE_PUSH_TO_TRACY option to DumpDeviceProfileResults

csehydrogen opened the pull request:
### Ticket
#20689 

### Problem description
Currently, there is no way to manually push dumped device profile results to tracy server. The results are pushed only when the process exits normally and dtor for the object holding data is properly called. If the process is stopped by a signal(for example, SIGINT by ^C) and there is no signal handler registered, data is lost.

### What's changed
This PR adds an option to trigger push.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

csehydrogen committed:
Add FORCE_PUSH_TO_TRACY option to DumpDeviceProfileResults

mo-tenstorrent commented:
I don't think we should do this. Post processing to do device device sync and device host sync, happen as part of the push. 
The device is also reordered to make it more suitable for how the GUI accepts the data. 

Why was this necessary? Do you want it for crashing runs that don't make it to the destructor?

csehydrogen commented:
Sorry for the lack of context.

To be more specific, we use tracy to profile model serving servers(vllm, sglang, our in-house framework, etc.). They usually run indefinitely and exit upon signals(SIGINT from user pressing ^C, SIGTERM, etc.). In most cases, they don't implement signal handlers, so dtor is not called and tracy server ends up having no device profile data.

Well, we can certainly put signal handlers to those, but having to implement signal handlers to make profiler work is kinda unconvincing. I thought more proper way is to call `DumpDeviceResults(FORCE_PUSH_TO_TRACY)` from the model code, for example, per user request.

I didn't understand the post processing part you described. As far as I understand, `pushTracyDeviceResults` is just grouping the results into `Device: id, Core (x, y)` contexts, so splitting the push should still result in the same timeline on GUI. I tested with this patch few times and I see no noticeable error in GUI.

csehydrogen commented on tt_metal/impl/profiler/tt_metal_profiler.cpp:
```diff
@@ -838,6 +838,9 @@ void DumpDeviceProfileResults(IDevice* device, std::vector<CoreCoord>& worker_co
                 InitDeviceProfiler(device);
             }
+            if (state == ProfilerDumpState::FORCE_PUSH_TO_TRACY) {
+                tt_metal_device_profiler_map.at(device_id).pushTracyDeviceResults();
```
I understood the situation. What about using TracyTTContextCalibrate when SYNC is ON, instead of assert? Seems like the function is made for this purpose.

csehydrogen replied:
SYNC is not a must-have, but it makes analysis much easier because the host timeline and device timeline aligns in GUI. So I want to support them if we can.

mo-tenstorrent replied:
I was not able to get Calibrate to work properly but honestly I did not spend that much time to figure out why. It would be nice to get it to work.

csehydrogen replied:
Ok, I'll take a look and make modifications either way.

csehydrogen committed:
Add profiler calibration

csehydrogen commented:
@mo-tenstorrent
Upon investigation, there were actually a few bugs.

* We need the following patch in tracy (which is external repository: https://github.com/tenstorrent-metal/tracy)
```
diff --git a/public/tracy/TracyTTDevice.hpp b/public/tracy/TracyTTDevice.hpp
index c6619833..4c87db6b 100644
--- a/public/tracy/TracyTTDevice.hpp
+++ b/public/tracy/TracyTTDevice.hpp
@@ -122,7 +122,7 @@ namespace tracy {
             MemWrite( &item->hdr.type, QueueType::GpuCalibration );
             MemWrite( &item->gpuCalibration.gpuTime, (int64_t)round((double)m_tgpu/m_frequency) );
             MemWrite( &item->gpuCalibration.cpuTime, tcpu );
-            MemWrite( &item->gpuCalibration.cpuDelta, mm_tcpu - tcpu );
+            MemWrite( &item->gpuCalibration.cpuDelta, (int64_t)((tcpu - mm_tcpu) * get_tracy_timer_mul()));
             MemWrite( &item->gpuCalibration.context, GetId() );
             Profiler::QueueSerialFinish();
```
Because we need to use wall-clock time, not tracy's scaled time. Also, the operand order for subtraction was wrong. (mm_tcpu - tcpu results in negative.)

* `syncDeviceHost` was using emplace to update `device_core_sync_info`, which **does not update** when the key already exists in the map.

* When `syncDeviceHost` computes cpuTime and delay via linear regression, it was using all samples(249 per sync). This makes delay, which is a device timestamp, sometimes go backward and breaks calibration. I modified it to use only the most recent sync info(exactly 249 samples).

Could you take a look, please? **Note that it should be merged after the aforementioned patch to tracy.**

csehydrogen commented:
I attached sample tracy files for all four cases: (push, sync) = (off, off), (off, on), (on, off), (on, on).
[tracies.zip](https://github.com/user-attachments/files/19811011/tracies.zip)

The below is the test script I used:
```
import torch
import ttnn
device = ttnn.open_device(device_id=0)
a = torch.rand(1, 1, dtype=torch.float32)
b = ttnn.from_torch(a, layout=ttnn.TILE_LAYOUT, device=device)
for j in range(3):
    for i in range(1000):
        ttnn.exp(b)
    ttnn.DumpDeviceProfiler(device)
ttnn.close_device(device)
```

mo-tenstorrent commented:
Great finds! The third item you mentioned is a needed feature. That is why we do sync every dump, to get long delays between sync clusters. That allows for a better linear regression estimation of Freq and delay.  Something else is bugged that is causing the sign flip you mentioned.

csehydrogen committed:
Regressions with all accumulated samples

ihnpark551 commented:
@mo-tenstorrent Hello, this is Ihn at Moreh.
Could you share the status of this item?
Because it is depending our Dev work, would like to ask your close support!

cc. @zzigler-tt @csehydrogen

csehydrogen commented:
Also made PR for tracy: https://github.com/tenstorrent-metal/tracy/pull/15

mo-tenstorrent commented:
Awesome fix! Let me just confirm a run on a multi-device run as well with device-device sync and force push on. I should have that ready before you start your day tomorrow. 

@ihnpark551 Unless the test above behaves unexpectedly, I believe we are good to merge this PR.

ihnpark551 commented:
@mo-tenstorrent,

Super Thanks for your reply! 

And also, could someone can be assigned to below PR? It is also related to this issue, which is for our Dev Work!
https://github.com/tenstorrent-metal/tracy/pull/15

mo-tenstorrent commented:
Unfortunately, the branch is failing on multi-device.

just apply
```
diff --git a/ttnn/cpp/pybind11/device.cpp b/ttnn/cpp/pybind11/device.cpp
index c70dac4e2c..3e0e489c2e 100644
--- a/ttnn/cpp/pybind11/device.cpp
+++ b/ttnn/cpp/pybind11/device.cpp
@@ -27,7 +27,8 @@ namespace py = pybind11;
 namespace {
 void DumpDeviceProfiler(IDevice* device) {
     ProfilerOptionalMetadata prof_metadata(tt::tt_metal::op_profiler::runtime_id_to_opname_.export_map());
-    tt::tt_metal::detail::DumpDeviceProfileResults(device, ProfilerDumpState::NORMAL, prof_metadata);
+    tt::tt_metal::detail::DumpDeviceProfileResults(device, ProfilerDumpState::FORCE_PUSH_TO_TRACY, prof_metadata);
+    //tt::tt_metal::detail::DumpDeviceProfileResults(device, ProfilerDumpState::NORMAL, prof_metadata);
 }
 }  // namespace
```

and run `pytest tests/ttnn/tracy/test_profiler_sync.py::test_all_devices`

You will see that device sync zones are broken and they disappear when you zoom in. Tracy does this with bad zones that have incomplete start and end pairs. 

Without the diff above, i.e. without mid-run force push the zones are fine. 

This is probably something simple, however our team is pretty backed up by more urgent tasks so I cannot look into it.

If this is getting outside the scope for you. I would go back to my original suggestion of not allowing mixing sync and mid-run force push.

mo-tenstorrent commented:
Note that we do not have the infra to test tracy GUI as part of CI. This is specially true for device side data.

This is why CI might be fine on this PR but we still require manual testing to make sure GUI is functional.

csehydrogen commented:
Thanks for confirmation. I indeed missed the multi-device tests. I think something is wrong with the SYNC functionality. Even without PUSH, the timelines of devices do not align up... I'll take a look at and estimate the amount of work we need to fix this. If too much, I'll just put assert when PUSH and SYNC is both on as you mentioned.

ihnpark551 commented:
Hello @mo-tenstorrent !
Thanks for your support! 
BTW, Could you assign adequate TT team member, for https://github.com/tenstorrent-metal/tracy/pull/15 ?
cc. @zzigler-tt @csehydrogen

csehydrogen committed:
Make multi-device profiling correct

csehydrogen commented:
@mo-tenstorrent I checked and fixed two main parts.

First, because `syncDeviceHost` is called for only one device (`first_connected_device_id`), `device_core_sync_info` is also filled in only one device and other devices have no idea about how to sync cpu time and device time. I modified `setSyncInfo` so that the sync information from that device propagates to the others.

Second, when doing linear regression between devices after `syncDeviceDevice` call, square of clocks are stored directly to double variables, which causes large numerical error. I added logic to shift clocks to the minimum value and run regression on shifted small values.

This is my traces with `test_profiler_sync.py` after this patch. It seems good, but I need your cross-validation.
[multidevice_traces.zip](https://github.com/user-attachments/files/19865656/multidevice_traces.zip)

ihnpark551 commented:
Hello @mo-tenstorrent and @zzigler-tt ,
@csehydrogen made fixes for this item,
and it would be great you to have a look of this item and [#15](https://github.com/tenstorrent-metal/tracy/pull/15) of tracy, as quick as possible!
(It is urgent to Moreh to have this item merged, because it is depends on Moreh's dev timeline.)


Thanks :) !

mo-tenstorrent commented:
@ihnpark551 I thought github will notify you. I assigned that PR to myself yesterday. No worries there, once we are satisfied with the state of things here we can merge that small PR.

mo-tenstorrent commented:
@csehydrogen 

Ok, first thing to point out is that main has sync accuracy on device-device in the range of 10s of ns. It is also functional on all devices.  So without force push that is the expectation and we need to deliver it. That is not what `multipushoffsyncon.tracy` is suggesting. Its sync accuracy is in milli seconds so something else is wrong.

Part of the confusion might be coming from this line [here](https://github.com/tenstorrent/tt-metal/blob/87ba548e59fadb32d6e4624ab0c24e13075da9f9/tt_metal/impl/profiler/profiler.cpp#L928) so basically we are syncing the first connected device with host and take that as the tracy sync for the entire cluster. We then re-align all the timestamps for each devices with the first connected device. This is only done for tracy atm but we are in the process of applying the same shift and scale to the raw timestamps sent to our csvs so we see synced timestamps across all sources of data. 

Note that we are not ignoring calculation errors either. In fact that is why we try to put sync time pair collections as far away as possible from each other.  Hence doing one at the beginning and one at the end. The further away, the bigger the deltas which brings sensitivity to math error down. 

In other words with device-device where propagation delay is very close to a constant(this is unlike device-host), 2 sync time pairs that are multiple seconds apart will yield to better accuracy than the 250 samples in one syncDeviceDevice round that are micro seconds apart. 

This is why a single device-device sync event is giving unusable sync values. Couple this fact with something that I just remembered during this review, which is that we cannot push eth kernels to Ethernet cores while fabric is up, and we are in a situation where we cannot do device-device sync without doing device close(i.e. fabric teardown) 

Put another way, a single device-device sync is not enough to get good sync parameters, and we cannot do anymore syncs while fabric is up. So device-device sync requires device close. 

My suggestion is that we turn force push into an environment variable. This actually has user experience benefits as well. We then only allow device-host sync when force push is asked and not attempt device-device sync. 

Long term, the goal is to bring sync time pairing as a feature to fabric so connected eths can sync their times without the need for pushing specific kernels to them and just by issuing a sync command to them. mid-run device-device sync can be revisited then.

csehydrogen committed:
Fix sync accuracy in multi-device profiling

csehydrogen commented:
@mo-tenstorrent Thanks for detailed review!

> Its sync accuracy is in milli seconds so something else is wrong.

It was my mistake that I forgot to multiply scale to receiverBase here. It is now accurate up to a few ns as main branch.

https://github.com/tenstorrent/tt-metal/blob/894d866ee06e6ed13d75997e503a3e480775beae/tt_metal/impl/profiler/tt_metal_profiler.cpp#L579-L580

> Note that we are not ignoring calculation errors either.

I understand that two device-device syncs that are multiple seconds apart make calibration more accurate, but still, having shifts (`senderBase` and `receiverBase` as above) results in much more accurate regression even with single sync. For example, if you try regression right after init (so, single device-device sync) in current main branch, it will produce these inaccurate freqScale and shift:

```
Device sync data for device: 7, delay: 12250946180 ns, freq scale: 0.9706790123456789
Device sync data for device: 3, delay: -13734495310 ns, freq scale: 1.0277777777777777
Device sync data for device: 4, delay: -42197499516 ns, freq scale: 1.0857142857142856
Device sync data for device: 6, delay: -102697413058 ns, freq scale: 1.211367942402425
Device sync data for device: 2, delay: -132878070357 ns, freq scale: 1.276847290640394
Device sync data for device: 5, delay: -100111435115 ns, freq scale: 1.206896551724138
Device sync data for device: 1, delay: -116174209706 ns, freq scale: 1.2413793103448276
Device sync data for device: 0, delay: 0 ns, freq scale: 1
```

However, with the new method, you get more accurate results:
```
Device sync data for device: 7, delay: 99985919 ns, freq scale: 0.9999930479382266
Device sync data for device: 3, delay: 106869796 ns, freq scale: 0.9999965658107491
Device sync data for device: 4, delay: 24078188 ns, freq scale: 1.0000053332026735
Device sync data for device: 6, delay: 120095892 ns, freq scale: 1.0000086245366815
Device sync data for device: 2, delay: 97130175 ns, freq scale: 1.000007085624121
Device sync data for device: 5, delay: 98225001 ns, freq scale: 1.0000058777642724
Device sync data for device: 1, delay: 93537714 ns, freq scale: 1.0000050896826531
Device sync data for device: 0, delay: 0 ns, freq scale: 1
```

Two syncs (one at init, one at close) make both more stable, but still, latter one is accurate.

---

Nevertheless, I reverted all changes for multi-device because multi-device is not our top priority and want this PR to be merged sooner. And I made it into `TT_METAL_PROFILER_FORCE_PUSH` envvar as you suggested. Hope this would be the final commit 😄 

But I really want you guys look at commit 894d866 in the future to make SYNC functionality work correctly on multi-device. In current main branch, only device 0 is synced with the host, as shown in the figure.

<img width="941" alt="Screenshot 2025-04-24 at 11 10 58 AM" src="https://github.com/user-attachments/assets/2264a62f-09b4-41cd-a028-bdb62499ec5c" />

mo-tenstorrent commented:
@csehydrogen just going through your device sync [commit](https://github.com/tenstorrent/tt-metal/commit/894d866ee06e6ed13d75997e503a3e480775beae). The performance is great and it is worth bringing it in. I am in the process of reverting it and testing it. There are couple of higher level tests and modification that I want to add which I should have ready soon.

mo-tenstorrent commented:
<img width="682" alt="Screenshot 2025-04-24 at 8 19 50 PM" src="https://github.com/user-attachments/assets/e11a5db0-6143-48fd-aa6c-9af9ec5f5748" />

So I am seeing worse cases such as the above were the sync error is much as 50us. This is only in for push mode.
In no-force-push mode we are 10ns accuracy. 
So I am gonna add a warning saying device-device sync is a best effort and should not be used for delicate device-device event analysis.

csehydrogen commented:
@mo-tenstorrent Hi,
Is it the case between device 0 and the other device? That is because, when PUSH is on, device 0 is calibrated with new syncDeviceHost results during DumpDeviceProfilerResults between open/close_devices, but other devices are still using old results. So, between other devices (for example, device 3 - 7), it will be still 10ns accuracy.

I try to come up with elegant solution, but because DumpDEviceProfilerResults accept a single device, not the whole mesh, I couldn't decide when to propagate syncDeviceHost result :(

So, for now, you're right; there should be a warning for using device-device sync and PUSH together

mo-tenstorrent commented:
Yes there is that and also the fact that older timestamps that have been push do not benefit from the second device sync that comes in because we cannot recalibrate them. 

Yeah this is good enough for now. Once we have fabric support for device-device sync we will revisit.

Also, I am basically dropping your last commit `4ebf5d2f`. Mainly the get sync info function was a little awkward with the statics and all. Hopefully, you don't mind.

mo-tenstorrent committed:
Reverting to mid-run device device sync

mo-tenstorrent committed:
Add unit test for mid-run tracy device push with sync

mo-tenstorrent commented on tt_metal/impl/profiler/tt_metal_profiler.cpp:
```diff
@@ -598,68 +679,13 @@ void ProfilerSync(ProfilerSyncState state) {
-            setSyncInfo(first_connected_device_id, (std::pair<double, int64_t>){1.0, 0}, deviceDeviceSyncInfo);
-        }
+    if (state == ProfilerSyncState::CLOSE_DEVICE and do_sync_on_close) {
+        do_sync_on_close = false;
```
@csehydrogen you sync host device here as well. There is a bug at the moment that does not allow that. For some reason we think we are in fast dispatch mode here and attempt to fast dispatch the kernels and fail. 

Removing it is fine though we already do device host syncs in dump device

mo-tenstorrent commented:
https://github.com/tenstorrent/tt-metal/actions/runs/14655738572
https://github.com/tenstorrent/tt-metal/actions/runs/14655741324

Once above pass, we should be good to merge.

ihnpark551 commented:
Huge appreciation for your fast support, mo!
And it would be great if we can have this PR merged till the end of this week! cc. @zzigler-tt

csehydrogen commented:
The test somehow passed on N300 but failed on N150:
https://github.com/tenstorrent/tt-metal/actions/runs/14655738572/job/41130599431

I cannot debug because I don't have access to N150. 😢 Could you check?

mo-tenstorrent commented:
[mid-run-push-samples-runs.zip](https://github.com/user-attachments/files/19912116/mid-run-push-samples-runs.zip)

Running: 

Just profiler regression from APC: https://github.com/tenstorrent/tt-metal/actions/runs/14668692078
T3K profiler: https://github.com/tenstorrent/tt-metal/actions/runs/14668701468

tt-rkim commented on tests/scripts/run_profiler_regressions.sh:
```diff
@@ -6,6 +6,14 @@ source scripts/tools_setup_common.sh
+    remove_default_log_locations
+    mkdir -p $PROFILER_ARTIFACTS_DIR
+    python -m tracy -v -r -p --sync-host-device --push-device-data-mid-run -m pytest tests/ttnn/tracy/test_profiler_sync.py::test_all_devices
+    runDate=$(ls $PROFILER_OUTPUT_DIR/)
```
Are we actually checking anything here?

mo-tenstorrent replied:
Just that a CSV is generated. This is basically a smoke test to make sure above options don't make the call hang

mo-tenstorrent replied:
Better commenting and naming would help?

mo-tenstorrent committed:
Fix N150 failure

abhullar-tt reviewed (approved):
changes to rtoptions lgtm
